File: python/qchallenge.py
import httplib, urllib, re, operator

operationDic = {"subtract": operator.sub, "add": operator.add, "divide": operator.div, "multiply": operator.mul}

# GET
http = urllib.urlopen("http://engineer.qamine.com/challenge")
response = http.read()

numbers = re.findall(r"\d+", response.split("\n")[0]) # find all numbers
numbers = map(lambda x : int(x), numbers) # convert them to int

operationType = response.split(" ")[4]

result = reduce(operationDic[operationType], numbers[:-1]) # apply operation to all number except the last one (id)

# POST
params = urllib.urlencode({'contact':'<>', 'payload': str(result), 'id': str(numbers[-1])})
http = urllib.urlopen("http://engineer.qamine.com/answer", params)
response = http.read()
print response

File: javascript/qchallenge.js
var http = new XMLHttpRequest();
http.open("GET", "/challenge", false);
http.send(null);

parts = http.responseText.split('\n')[0].split(' ')

a = parseInt(parts[5])
b = parseInt(parts[7])

if (parts[4] == 'subtract') result = b-a
if (parts[4] == 'add') result = a+b
if (parts[4] == 'divide') result = a/b
if (parts[4] == 'multiply') result = a*b

result = parseInt(result)

params = 'contact=<email.address>&payload='+result+'&id='+parts[15]

http.open('POST', '/answer', true);
http.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
http.send(params);

http.onreadystatechange = function() {
	if(http.readyState == 4 && http.status == 200) {
		console.log(http.responseText);
	}
}
